config/mysqlpersister/mysqlpersister.go

```diff
@@ -0,0 +1,215 @@
+	shutdown        chan struct{}
+	fetcherShutdown chan struct{}
+
+	configs map[int]*qsc.ServiceConfig
+}
+
+type configRow struct {
+	Version int    `db:"Version"`
+	Config  string `db:"Config"`
+}
+
+type Connector interface {
+	Connect() (*sql.DB, error)
+}
+
+func New(c Connector, pollingInterval time.Duration) (*MysqlPersister, error) {
+	db, err := c.Connect()
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = db.Exec("SELECT 1 FROM quotaservice LIMIT 1")
+	if err != nil {
+		return nil, errors.New("table quotaservice does not exist")
+	}
+
+	mp := &MysqlPersister{
+		db:              db,
+		configs:         make(map[int]*qsc.ServiceConfig),
+		m:               &sync.RWMutex{},
+		watcher:         make(chan struct{}),
+		shutdown:        make(chan struct{}),
+		fetcherShutdown: make(chan struct{}),
+		latestVersion:   -1,
+	}
+
+	if _, err := mp.pullConfigs(); err != nil {
+		return nil, err
+	}
+
+	go mp.configFetcher(pollingInterval)
+
+	return mp, nil
+}
+
+func (mp *MysqlPersister) configFetcher(pollingInterval time.Duration) {
+	defer func() {
+		close(mp.fetcherShutdown)
+	}()
+
+	for {
+		select {
+		case <-time.After(pollingInterval):
+			if newConf, err := mp.pullConfigs(); err != nil {
+				logging.Printf("Received an error trying to fetch config updates: %s", err)
+			} else if newConf {
+				mp.notifyWatcher()
+			}
+		case <-mp.shutdown:
+			logging.Print("Received shutdown signal, shutting down mysql watcher")
+			return
+		}
+	}
+}
+
+// pullConfigs checks the database for new configs and returns true if there is a new config
+func (mp *MysqlPersister) pullConfigs() (bool, error) {
+	mp.m.RLock()
+	v := mp.latestVersion
+	mp.m.RUnlock()
+
+	q, args, err := sq.
+		Select("Version", "Config").
+		From("quotaservice").
+		Where("Version > ?", v).
+		OrderBy("Version ASC").ToSql()
+	if err != nil {
+		return false, err
+	}
+
+	rows, err := mp.db.Query(q, args...)
+	if err != nil {
+		return false, err
+	}
+
+	rowCount := 0
+	maxVersion := -1
+	for rows.Next() {
+		rowCount++
+
+		var r configRow
+		err := rows.Scan(&r.Version, &r.Config)
+		if err != nil {
+			return false, err
+		}
+
+		var c qsc.ServiceConfig
+		err = proto.Unmarshal([]byte(r.Config), &c)
+		if err != nil {
+			logging.Printf("Could not unmarshal config version %v, error: %s", r.Version, err)
+			continue
+		}
+
+		mp.m.Lock()
+		mp.configs[r.Version] = &c
+		mp.m.Unlock()
+
+		maxVersion = r.Version
+	}
+
+	if rowCount == 0 {
+		return false, nil
+	}
+
+	mp.m.Lock()
+	mp.latestVersion = maxVersion
+	mp.m.Unlock()
+
+	return true, nil
+}
+
+func (mp *MysqlPersister) notifyWatcher() {
+	mp.watcher <- struct{}{}
+}
+
+// PersistAndNotify persists a marshalled configuration passed in.
+func (mp *MysqlPersister) PersistAndNotify(_ string, c *qsc.ServiceConfig) error {
+	b, err := proto.Marshal(c)
+	q, args, err := sq.Insert("quotaservice").Columns("Version", "Config").Values(c.GetVersion(), string(b)).ToSql()
+	if err != nil {
+		return err
```

Comment:
In the event that `PersistAndNotify` is called with a version that already exists in the table, I'm wondering if we should return a typed error. If the error was typed, it would be easier for the calling code to act appropriately on that duplicate version.

Not a PR blocker, just a thought. The packages may already return a typed error.

Reply:
That's a really good call.  I added it, that way they don't have to deal with parsing the mysql error to figure out if it was caused by a dup (better abstraction hopefully?)